### PR TITLE
Check for space in buffer before writing tiles

### PR DESCRIPTION
**Pull request by matt335672**
This is a required part of the fix for neutrinolabs/xrdp#2068

Thanks to @trishume for supplying this patch.

### CODE REVIEW

**Comment by matt335672** [comment]
@trishume - I had a question for you which you may be able to answer fairly swiftly.

in `check_and_rfx_encode()`, why is your overflow size check value set to 8192 rather than the uncompressed tile size of 16384?

Thanks.

**Comment by trishume** [comment]
@matt335672 It's been over a year since I wrote that so I don't really remember, but my vague recollection is I was thinking like "I'll assume the compression won't do more than double the uncompressed size even in a pathological case" and that's where the *2 comes from. And then perhaps I was just wrong about the uncompressed size and thought it was 4096.

File: src/rfxencode_tile.c [hunk 1]
```diff
@@ -97,6 +97,19 @@ rfx_encode_component_rlgr3(struct rfxencode *enc, const char *qtable,
     return 0;
 }
 
+/******************************************************************************/
+static int
+check_and_rfx_encode(struct rfxencode *enc, const char *qtable,
+                     const uint8 *data,
+                     uint8 *buffer, int buffer_size, int *size)
+{
+    if (buffer_size < 4096 * 2)
```
Comment (matt335672) [inline src/rfxencode_tile.c]: This magic number needs explaining - it's not clear where it comes from

**Comment by matt335672** [comment]
@trishume - I agree that your general approach to assume the compression won't do worse than double the uncompressed size is sound. Someone who really understands the maths behind the compression might be able to come up with a lower bound than that, but that person isn't me.

I've used [MS-RDPRFX] to come up with a number, and I've tried to express the logic in comments and macro definitions.

The value I've ended up with is bigger than your value of 8192, but even so, I could gen anywhere near 8192 in practice - I managed 5774. Maybe a cunningly crafted packet could break 8192, but I've not idea how.

**Review by metalefty — APPROVED** [review]

**Pull request merged**